### PR TITLE
Change Ref<T> to allow non const access to ptr

### PR DESCRIPTION
**Pull request by derammo**
# Summary

This PR changes `Ref<T>` to allow non-const access to the referenced object even from a `const Ref<T>`.

Exactly as in `std::shared_ptr<T>`, the constness of the `Ref<T>` refers only to whether the pointer can be changed to point to something else.  The actual underlying object is not const because the pointer being managed is a non-const `T*`.  Due to restrictions with `RefCounted` (internal ref count), we cannot compile `Ref<const T>`, but that is ok because Godot is permissive when in doubt.  In Godot, a const object can just be passed as a `const T&` or `const T*` when this is truly needed.  

In practice, the following methods (and possibly more) already allowed non-const access to a `const Ref<T>`:

- `Ref<T>::operator=(const Ref<T>&)` turns a `const Ref` into a non-const `Ref`
- `Ref<T>(const Ref<T>&)` turns a `const Ref` into a non-const `Ref`
- `Ref<T>(const Variant&)` turns a `const Variant` containing a ref into a non-const `Ref`
- `Ref<T>::operator Variant() const` turns a `const Ref` into a Variant containing non-const `Ref`

# Problem being solved

In functions receiving a `const Ref<T>&`, the T object cannot be called on non-const methods.  That means it is frequently not usable by the function to which it is being passed.

This is shown in https://github.com/godotengine/godot-cpp/issues/693

It happens a lot on the Extension side because `const CLASS&` is the way the binding generator writes arguments and that is `const Ref<T>&` for reference counted types.

# godot-cpp

If this can be approved, a companion PR will make the equivalent changes in `godot-cpp`.

### CODE REVIEW

**Comment by Bromeon** [comment]
`std::shared_ptr` provides const-correctness through the CV qualification of its template argument: `shared_ptr<const T>`.
However, this doesn't seem to be a convention that Godot uses (yet?)

Relying on top-level CV qualifiers for const-correctness doesn't really work:
```cpp
void function(const Ref<T>& ref) {
    Ref<T> copy = ref;
    copy->mutating_method();
}
```

What's the canonical way in Godot (and godot-cpp) to express "here's a reference to `x` but I don't want you to modify `x`"?

**Comment by derammo** [comment]
I believe I addressed your comments above already.

> Due to restrictions with `RefCounted` (internal ref count), we cannot compile `Ref<const T>`

The only thing that is truly const is `const T&` and if you take a `const T&` where `T` is `RefCounted` then you should not be allowed to make `Ref<T>` from it.   Because otherwise the code receiving this reference could end up holding the last ref count and destroy the object later on.  That's pretty much never what you wanted.   I use `const T&` to mean you can use const methods on this and you can't retain it.   There isn't any written convention I think, but `godot-cpp` is emitting `const &` for const anything at the moment, which is how we got here.

We could in theory make ref counts mutable and allow something like a `Ref<const RefCounted>` but I think that is some academic stuff that helps nobody get work done.   You don't want me to call methods on this but its ok if I delete it later?  What's the use case there?  As I understand it, Godot is permissive to get work done and right now `Ref<>` is being unhelpful for no good reason.

**Comment by groud** [comment]
We reviewed the PR in the GDextension meeting and came to the conclusion this is not the right fix to the issue. The Ref<T> "const-ness" works fine on the Godot side and the issue should be solved on the godot-cpp side.

Several solutions could be envisioned:
- 1) add some way to register a parameter "constness" to ClassDB, which would allow to add the information to `extension_api.json`, and thus generate the godot-cpp code taking this information into account. This might however be difficult to do as it would probably need a huge pass over the codebase.
- 2) generate the godot-cpp code with const qualifiers on arguments only if the function itself is const-qualified.  That should cover most situations. This might be the cleanest solution for now.
- 3) remove the const qualifiers on the godot-cpp side only.

For now we close this PR as we don't think this is the right solution to the issue, thank you nonetheless. Feel free to open PR with one of the solutions above, likely the second one would be preferred.

**Comment by Faless** [comment]
I'm re-opening after further discussion in the gdextension chat.

I think @derammo is correct on this:

> Exactly as in std::shared_ptr<T>, the constness of the Ref<T> refers only to whether the pointer can be changed to point to something else

and as his and @Bromeon 's remarks show, we have a lot of cases where in fact `const Ref<T>` is converted to `Ref<T>` and then `T` used as non-const.

As to how to achieve constness in Godot, I don't think it is currently possible if the scope it for it to cross the c++ boundaries (i.e. be exposed).

We might be able to express the concept for virtual methods, but doing so for other exposed methods might be trickier (because variant does not support it).

In any case, I think for now we should:
- Recognize the fact that `const Ref<T> != Ref<const T>`
- Embrace the fact that godot only supports `const Ref<T>` (and `Ref<T>` for returns).
- Merge this PR.
- Apply the same changes to `Ref` to `godot-cpp` as suggested by @derammo .

**Review by touilleMan — APPROVED** [review]

**Comment by saki7** [comment]
cc: @dsnopek 
I've noticed that this issue has been discussed on the recent GDExtension meeting (2023-05-13).

[The meeting log](https://docs.google.com/document/d/1B1ffNeNX5u24C8Oh3SF2vRrnEsurrithrFBgUZRHHCo/edit#) says: 

> - Need a good use case for why this is needed

I guess the team is still not sure why this PR needs to be merged.

## My opinion
@Faless's [comment](https://github.com/godotengine/godot/pull/64789#issuecomment-1353271224) is correct. Please merge this PR.


## Why is this PR correct?
Long story short: It boils down to what  @Faless said:
> - Recognize the fact that `const Ref<T> != Ref<const T>`


If it's permitted to add further rationale for this point, I would say:
- C++ programmers are used to C++ language's semantics of pointer variables.
- In the C++ standard library, smart pointers are *designed* to have the semantics exactly the same as raw pointers.
  - See: <https://en.cppreference.com/w/cpp/memory/shared_ptr/operator*>
  - const qualifier for pointer type designates the const-ness for the raw pointer variable itself.
  - const qualifier for element type designates the const-ness for the underlying value which the raw pointer points to.
- C++ programmers normally expect that an arbitrary smart pointer types provided by an external framework do imply the similar semantics as the standard smart pointers. Otherwise, it will look *strange*, thus leading to surprising errors.
- Technically speaking, Godot's `Ref<T>` is one form of a smart pointer class.
- I couldn't find a technical reason why Godot's `Ref<T>` *must* behave differently than the standard smart pointers.
- This PR should be merged to ensure the semantics will be equivalent to the standard smart pointers.


## Realistic use-case
If you understood the above description, you don't need to read this section. If you're still confused, please take your time to read this example.



```cpp
struct EmployeeAttributes
{
    String name;
    unsigned age = 0;
    bool operator==(const EmployeeAttributes&) const noexcept = default;
};

class EmployeeData : public RefCounted
{
public:
    GDCLASS(EmployeeData, RefCounted);
    static void _bind_methods() {}

    std::int64_t id;
    std::optional<EmployeeAttributes> attrs;
};


class EmployeeNameCard : public PanelContainer
{
public:
    GDCLASS(EmployeeNameCard, PanelContainer);
    static void _bind_methods() {}

    Ref<EmployeeData> get_data() const { return data_; }
    void set_data(Ref<EmployeeData> data);

private:
    Ref<EmployeeData> data_;
};


// Updates the employee data for this card.
// Also assigns default attrs to `new_data` if non-null attrs are required for displaying GUI.
// Caller can compare the resulting attrs of `new_data` with old data to check if any change was made.
void EmployeeNameCard::set_data(Ref<EmployeeData> new_data) // <===========  NOTE HERE [2]
{
    if (!new_data->attrs) {
        // Assign default values if new_data.attrs is empty.
        // This is required because we need non-null info for displaying GUI anyways
        new_data->attrs = EmployeeAttributes{};  // <===========  NOTE HERE [1]
    }

    data_ = new_data;
}


void update_name_card_by_user_input_data(EmployeeNameCard* card, Ref<EmployeeData> new_data)
{
    auto old_data = card->get_data();
    card->set_data(new_data);

    if (*old_data->attrs != *new_data->attrs) {  // <===========  NOTE HERE [4]
        std::cerr << "TODO: GUI card has been updated, we need to update corresponding DB too" << std::endl;
    }
}
```


First, take a look at [1].

```cpp
 new_data->attrs = EmployeeAttributes{};  // <===========  NOTE HERE [1]
```

This compiles, because `new_data` was NOT const-qualified, as in [2]:

```cpp
  void EmployeeNameCard::set_data(Ref<EmployeeData> new_data) // <===========  NOTE HERE [2]
```

But soon you start to wonder: 
*"What if someone accidentally re-assigns something to `new_data`?"*

```cpp
void EmployeeNameCard::set_data(Ref<EmployeeData> new_data)
{
    if (new_data.is_null()) {
        new_data = db->get_dummy_data<EmployeeData>();
    }

   // original code follows...
```

The change above will lead to unexpected behavior because, even when you have new variable re-assigned to `new_data`, the `new_data` variable on caller side still points to the old variable [4]:

```cpp
    if (*old_data->attrs != *new_data->attrs) {  // <===========  NOTE HERE [4]
        std::cerr << "TODO: GUI card has been updated, we need to update corresponding DB too" << std::endl;
    }
```

In this kind of situation, C++ programmers tend to put additional const qualifier to the formal parameter, for the sake of sanity. Note that changing parameter to non-const lvalue reference is a wrong solution here, because if you do so it will become impossible to call `set_data` with prvalue.

So you innocently add `const` to [2]:

```cpp
void EmployeeNameCard::set_data(const Ref<EmployeeData> new_data) // `const` is added
{
```


If the smart pointer class, i.e. `Ref<T>`, was designed correctly, your job is done. However, the change above will lead to an unexpected error as of now. See the original code [1]:

```cpp
        new_data->attrs = EmployeeAttributes{};  // <===========  NOTE HERE [1]
```

Above code will start to fail because the current implementation of `Ref<T>::operator->` returns overly-const-qualified value. It doesn't match the usual semantics for pointers, and there's no viable workaround on user-side. The only solution for this is to fix the design of `Ref<T>`.

**Comment by dsnopek** [comment]
@saki7 Thanks so much for the additional information!

I need to spend some more time looking into this, but here's my (personal) first thoughts.

It sounds like a big part of the motivation for this change is that `Ref<T>` in Godot treats const-ness differently than the smart pointers from the C++ standard library. Here's a little table (which I hope is correct):

| Raw pointers | Godot currently | C++ standard library |
| --- | --- | --- |
| `T *foo` | `Ref<T> foo` | `std::shared_ptr<T> foo` |
| `const T *foo` | `const Ref<T> foo` | `std::shared_ptr<const T> foo` |
| `T* const foo` | `Ref<const T> foo` (not sure anyone has ever used this) | `const std::shared_ptr<T> foo` |

Am I understanding correctly that the goal here is to make:

- `const Ref<T> foo` become equivalent to `T* const foo` and
- `Ref<T const> foo` become equivalent to `const T *foo`

... to match how `std::shared_ptr` works?

Looking at the Godot codebase, it's very consistently using `const Ref<T> foo` to be equivalent to `const T* foo`, where the actual value that is being pointed at is const and shouldn't be changed.

So, if we were to make this change, it sounds like we'd have to change every instance of `const Ref<T>` in Godot to be `Ref<const T>` to keep the same meaning? Is that correct?

Apologies if I'm misunderstanding. I'm not super familiar with the smart pointers from the C++ standard library, and I haven't spent very much time digging into this yet.

Thanks again for taking the time to make such a detailed explanation!

**Comment by saki7** [comment]
@dsnopek: It seems to me that your understanding's basically correct, but there are actually two stages of mistakes on Godot (and godot-cpp):

- The first mistake was the design flaw on `Ref<T>`. It will be corrected when this PR is merged.
- The secondary mistake was the incorrect materialization of `const T* foo` semantics.

Speaking in a pedantic fashion, one must use either `YourSmartPointer<const T>` or `const T*` for the second case. However, like Faless said in the comment, Godot core and godot-cpp historically utilized the form `Ref<T>` (and/or `const Ref<T>`) for those cases. This is semantically wrong, thus it could have lead to incorrect extraction of mutable value:

>  (https://github.com/godotengine/godot/pull/64789#issuecomment-1353271224)
> ... lot of cases where in fact `const Ref<T>` is converted to `Ref<T>` and then T used as non-const.

However, this kind of usage should already have spread all across the repo, so it will be a painful and long road to fix the all occurrences. To be honest I don't even want to try to suggest this kind of changes right now, and it's also out of the scope of this PR.

Anyway, we should correct the `const Ref<T>` semantics in the first place; otherwise there's no point of fixing the wrong pointer semantics based on a wrong smart pointer design. And this is exactly the benefit of merging this PR.

All this explanation above could be summarized into the points Faless mentioned:

> (https://github.com/godotengine/godot/pull/64789#issuecomment-1353271224)
>
> In any case, I think for now we should:
>
> - Recognize the fact that `const Ref<T> != Ref<const T>`
> - Embrace the fact that godot only supports `const Ref<T>` (and `Ref<T>` for returns).
> - Merge this PR.
> - Apply the same changes to `Ref` to `godot-cpp` as suggested by @derammo .

**Comment by dsnopek** [comment]
Ok, I think I'm understanding this a little more.

Fundamentally, Godot has been trying to use `const Ref<T> foo` to be equivalent to `const T* foo`, however:

1. **It really isn't** because you can simply copy to a non-const `Ref<T>` and change the thing it's pointing to (stealing @Bromeon's example above):
    ```
    void function(const Ref<T>& ref) {
        Ref<T> copy = ref;
        copy->mutating_method();
    }
    ```
2. **It can't be** because (as far as I know - I could be wrong on this point) C++ will always allow you to copy a variable that's `const` to a new variable that isn't. So, there is no (again, as far as I know) C++ magic that could actually enforce `const Ref<T>` being truly equivalent to `const T*`, even though we've been using it as if it was equivalent (except when we needed to cheat :-))

Of course, const-ness has always been sort of an illusion in C++: in most cases, you can explicitly cast your way out of `const`, but the compiler is supposed to put up a fight when you try to violate const-ness without doing an ugly explicit cast, and with how `const Ref<T>` works, it doesn't put up any fight at all (and I don't think we could make it do so, per point nr 2 above).

Another signal that Godot's `Ref<T>` isn't architected quite right, is that we don't have any way to make a reference that's equivalent to `T* const` (where the pointer itself can't be modified, but the thing that's being pointed to can). This isn't something that we'd probably need very often (or maybe even at all, since we haven't had it all this time :-)), but the fact that it's not possible points to a hole in `Ref<T>`'s design.

So, I think I've personally come around to thinking this change is correct.

However, this is a change to core, that has wide-reaching impact through out the engine, and on how everyone writes code for it (ie. we should probably start writing `Ref<const T>` in those situations where we truly wanted an equivalent to `const T*`). This is the sort of decision that should be made by the @godotengine/core team, and no member of that team has reviewed this yet.

**Comment by derammo** [comment]
> However, this is a change to core, that has wide-reaching impact through out the engine, and on how everyone writes code for it (ie. should we start writing `Ref<const T>` in those situations where we truly wanted an equivalent to `const T*`?). This is the sort of decision that should be made by the @godotengine/core team, and no member of that team has reviewed this yet.

@dsnopek almost.  I realize this thread is rather long now and I may have used too short an explanation in the original message.  But keep in mind that you **cannot** currently use `Ref<const T>`.   This is because `Ref` uses an "internal ref count" implementation.    A fine choice to save heap objects, of course.  It just means you can't separate the const-ness of the ref count from the const-ness of the object.   Don't try to fix that, or you end up in a situation where you make the ref count` `mutable`, and then you can delete an object through a const reference, which surely you didn't want. :)

That's why in my original message I suggested to just use `const Ref<T>` for what it actually means in C++ (a const reference to a non-const thing, like `T* const`).   Then you can write a function like `void foo(const Ref<T>& shared)` which treats Ref like any other object, for example in a templated method, or `void foo(Ref<T> shared)` to use it like a value and incur the extra AddRef/Release pair.  Either way, you get the results that are normal for C++.  

If you really want to write code that passes a reference counted object const (so you can't even create a temporary additional reference to it) then you would have to pass it as a naked `const T*`.   But since Godot doesn't really do const-ness, this isn't necessary in any cases I can imagine. 

In case that wasn't clear, you can't just do what standard C++ shared_ptr does to implement `Ref<const T>`.   It uses a separate object for the ref count, which means it CAN have separate const-ness of the reference vs the object.   Obviously that would be a terrible idea for Godot :)

**Comment by derammo** [comment]
> Of course, const-ness has always been sort of an illusion in C++: in most cases, you can explicitly cast your way out of `const`, but the compiler is supposed to put up a fight when you try to violate const-ness without doing an ugly explicit cast, and with how `const Ref<T>` works, it doesn't put up any fight at all (and I don't think we could make it do so, per point nr 2 above).

You say "illusion" and I say "extremely valuable static type checking feature that detects errors early and saves countless development dollars" but ok :)

**Comment by derammo** [comment]
> * I couldn't find a technical reason why Godot's `Ref<T>` _must_ behave differently than the standard smart pointers.

@saki7 see my previous message for the technical reason it has to be different.  However, the parts that it CAN implement, such as `operator *` and `operator ->` could and should be done the same way, at least in terms of const-ness.   I know we can't fix the `operator*` returning the "wrong" type because of the many core changes that would entail.  That's why I didn't suggest that change.  But at least the const-ness can be fixed, making it usable in normal scenarios (a function receives a const reference to a temporary Ref and invokes methods on it.)

**Comment by saki7** [comment]
@derammo Thanks for the clarification :)  I have no objection for keeping current design for `operator*`.

For now we can merge this PR to fix merely these three methods, `Ref<T>::operator->`, `Ref<T>::operator*`, `Ref<T>::ptr()`. 

Like @derammo, I also don't want to suggest changing all parts of core to use `Ref<const T>` because (1) that's going to be a huge change, and (2) it's out of the scope of this PR.

I ensure this PR will not introduce any regression, since the new implementation will always return the mutable pointer (which is correct in terms of const-ness) and the core had already been working on the mutable value. This point has also been confirmed by multiple contributors on this thread.

I think that we now understood what the real problem is, and the solution have been validated by multiple people in this thread. I believe the last concern raised by GDExtension team is now resolved.

Now I strongly seek someone with leadership to decide whether to merge this PR.

**Comment by dsnopek** [comment]
Re @derammo:

>> However, this is a change to core, that has wide-reaching impact through out the engine, and on how everyone writes code for it (ie. should we start writing Ref<const T> in those situations where we truly wanted an equivalent to const T*?)
> almost. I realize this thread is rather long now and I may have used too short an explanation in the original message. But keep in mind that you cannot currently use Ref<const T>. This is because RefCounted uses an "internal ref count" implementation. A fine choice to save heap objects, of course. It just means you can't separate the const-ness of the ref count from the const-ness of the object.

Ah, ok. I think I'm really closing in on fully understanding this. :-)

To summarize really, really briefly: We been sort of pretending that `const Ref<T>` is like `const T*`, but it really isn't and it can't be (for the reasons you list above - the reasons in my previous comment were a little off), and so, we should stop pretending, and adjust the implementation to remove the bits that are (unsuccessfully) attempting to enforce the illusion.

From this perspective, we don't need to change how we're using `const Ref<T>` because we've really been using it this way all along, we're just harmonizing the implementation with itself.

Is that correct?

Re @saki7:

> I think that we now understood what the real problem is, and the solution have been validated by multiple people in this thread. I believe the last concern raised by GDExtension team is now resolved.

Well, I'm personally convinced about this PR now, but I can't speak for the whole GDExtension team. :-)

>Now I strongly seek someone with leadership to decide whether to merge this PR.

I agree that this is the next step! In the end, since this is core, it is really up to the @godotengine/core team to decide.

**Comment by derammo** [comment]
...
> To summarize really, really briefly: We been sort of pretending that `const Ref<T>` is like `const T*`, but it really isn't and it can't be (for the reasons you list above - the reasons in my previous comment were a little off), and so, we should stop pretending, and adjust the implementation to remove the bits that are (unsuccessfully) attempting to enforce the illusion.
> 
> From this perspective, we don't need to change how we're using `const Ref<T>` because we've really been using it this way all along, we're just harmonizing the implementation with itself.
> 
> Is that correct?
...

Yes.  It actually is like `T* const` but some inconsistent restrictions were placed on the `T` to try to make it act similar to `const T*` in some, but not all, cases.   As a result, any code that expects it to act like a normal C++ object fails.  In this case, it is primarily code on the Godot-CPP side where code is emitted like `const Ref<T>&` like you would normally pass an object, and then you can no longer call functions on the T without brutally removing the const-ness.

You are correct that what was attempted with these restrictions never worked.  It never did much useful work, only caused problems.   The implementation of `Ref` needs to be consistent between core and Godot-CPP so that code remains portable across.   That's really the only reason this Core PR exists, because it can only be changed in BOTH places and Core is the harder place to get approval.   

Don't forget to change it in Godot-CPP if you go forward!   While existing code should not be affected, code written after the change may start to use it without the extra casting/conversion so it will start requiring the changes in this PR.

**Comment by derammo** [comment]
I am sorry to introduce a new thought when this is almost converged.  So skip this if you want :).

But I just realized you COULD if you wanted introduce a separate object like `ConstRef<T>` if people in Core really really want something that is analogous to `const T*`.   Since `Ref<const T>` can't be compiled (as discussed above), you could have such a thing that represents "a mutable reference to an immutable object" which is what `const T*` is.

**Review by reduz — APPROVED** [review]

**Pull request merged**
